soufi/finders/yum.py

```diff
@@ -0,0 +1,215 @@
+# Copyright (c) 2021 Cisco Systems, Inc. and its affiliates
+# All rights reserved.
+
+import abc
+import functools
+import urllib
+
+import repomd
+import requests
+
+from soufi import exceptions, finder
+
+TIMEOUT = 30  # seconds
+
+
+class YumFinder(finder.SourceFinder, metaclass=abc.ABCMeta):
+    """An abstract base class for making Yum-based finders.
+
+    Subclasses of this should provide methods for setting up the repository
+    search lists to pass to __init__.
+
+    The lookup is a 2-stage process:
+        1. Attempt to look up the SRPM directly from the sources repository.
+           In these cases, we use the name only and ignore the version,
+           since it is a very common practice for repositories to only
+           publish repodata for the "current" version, while still
+           providing older sources.
+        2. Attempt to look up the name of the source RPM from the binary
+           repository.  This will catch packages where the source and
+           binary package names do not match.  The version is also ignored
+           in this step, for the same reasons.  Then backtrack and attempt
+           to lookup that package in the source repos.
+    """
+
+    def __init__(self, *args, source_repos=None, binary_repos=None, **kwargs):
+        self.source_repos = source_repos
+        if not self.source_repos:
+            self.source_repos = self._get_source_repos()
```

Comment:
It feels like a lot of complexity can be removed if you get rid of the source_repos and binary_repos init args here, and rely solely on the abstract get_X_repos() methods to provide the list/generator when requested. You can them remove the stuff in the derived finder that is attempting to re-initialise the generator.

Reply:
If I'm understanding this ask correctly (hint: it's early here so likely I'm not) that would remove the ability to instantiate the finder using private/local/etc. mirrors, which is absolutely something that I want to be able to do.

Reply:
You can still wrap that inside the get method I think. Just take it out of the base class. It's the direct cause of the confusion and need to reset the generator with a flag.

Which brings me to another thought - if someone passes a private list then it won't act as a generator at all, so I think this whole area needs some redesigning to wrap those in generators.

Reply:
I *just** moved this logic into the base class to get rid of all the lookalike `__init__` methods in all the derived classes, now you want me to put all that back?

Reply:
That was before you added the generator! You can still wrap everything in a generator in this __init__ though if you want.  I just dislike the flag that says to restart the generator, it seems unnecessary if you just make everything a generator.